source/applications/advanced/color_balance.py

```diff
@@ -0,0 +1,257 @@
+        pixels: Number of central pixels for computation
+
+    Returns:
+        mean_red: Red channel mean value
+        mean_green: Green channel mean value
+        mean_blue: Blue channel mean value
+
+    """
+    height = np.shape(rgb)[0]
+    width = np.shape(rgb)[1]
+    red = rgb[
+        int((height - pixels) / 2) : int((height + pixels) / 2),
+        int((width - pixels) / 2) : int((width + pixels) / 2),
+        0,
+    ]
+    green = rgb[
+        int((height - pixels) / 2) : int((height + pixels) / 2),
+        int((width - pixels) / 2) : int((width + pixels) / 2),
+        1,
+    ]
+    blue = rgb[
+        int((height - pixels) / 2) : int((height + pixels) / 2),
+        int((width - pixels) / 2) : int((width + pixels) / 2),
+        2,
+    ]
+
+    mean_red = np.mean(np.reshape(red, -1, []))
+    mean_green = np.mean(np.reshape(green, -1, []))
+    mean_blue = np.mean(np.reshape(blue, -1, []))
+
+    return mean_red, mean_green, mean_blue
+
+
+def _rgb_from_3d_capture(camera):
+    """Get RGB image from 3D capture.
+
+    Args:
+        camera: Zivid camera
+
+    Returns:
+        rgb: RGB image
+
+    """
+    frame = camera.capture()
+    point_cloud = frame.get_point_cloud().to_array()
+    rgb = np.dstack([point_cloud["r"], point_cloud["g"], point_cloud["b"]])
+
+    return rgb
+
+
+def _color_balance_calibration(camera):
+    """Balance color for RGB image by taking images of white surface (piece of paper, wall, etc.) in a loop.
+
+    Args:
+        camera: Zivid camera
+
+    Returns:
+        corrected_red_balance: Corrected red balance
+        corrected_blue_balance: Corrected blue balance
+
+    """
+    print(f"Starting color balance calibration")
+    default_red_balance = camera.settings.red_balance
+    default_blue_balance = camera.settings.blue_balance
+    with camera.update_settings() as updater:
+        updater.settings.blue_balance = 1.0
+        updater.settings.red_balance = 1.0
+
+    rgb = _rgb_from_3d_capture(camera)
+    [mean_red, mean_green, mean_blue] = _compute_mean_rgb(rgb, 100)
+    print(
+        f"Mean color values: R =",
+        int(mean_red),
+        f"G =",
+        int(mean_green),
+        f"B =",
+        int(mean_blue),
+    )
+    _display_rgb(rgb, "RGB image before color balance (3D)")
+
+    while int(mean_green) != int(mean_red) and int(mean_green) != int(mean_blue):
+        corrected_red_balance = camera.settings.red_balance * mean_green / mean_red
+        corrected_blue_balance = camera.settings.blue_balance * mean_green / mean_blue
+        with camera.update_settings() as updater:
+            updater.settings.red_balance = corrected_red_balance
+            updater.settings.blue_balance = corrected_blue_balance
+        rgb = _rgb_from_3d_capture(camera)
+        [mean_red, mean_green, mean_blue] = _compute_mean_rgb(rgb, 100)
+        _display_rgb(rgb, "RGB image after color balance (3D)")
+
+    return (
+        corrected_red_balance / default_red_balance,
```

Comment:
This can't be correct? Must be multipy?

Reply:
It is correct in my opinion (and it is proved to work). The point is that this gain is applied to an image here: https://github.com/zivid/python-samples/blob/color-balance/source/applications/advanced/color_balance.py#L212
That image is a result of capture with default_red_balance. That is why you have to divide it with default_red_balance to get rid of it and multiply with the corrected_red_balance because that is what you want.

Maybe it would be more understandable if this function was to return just corrected_red_balance, and then when we apply it we divide by default_red_balance and multiply wiht the corrected_red_balance. However, then we need to save the defaut_red_balance in main and transfer it to _apply_color_balance, or hard code it.

What do you think?

Reply:
Ok, so since you can't turn of Rgain and Bgain for 2D capture. You have to multiply image with the corrected settings. While in 3D then new Rgain is Rgain*correctedRgain. Not sure that's the easiest format for it.